Explain ## identifiers in manual

lassik opened the pull request:
Lightly edited from [Marc's comment](https://github.com/gambit/gambit/commit/e05d171759c9453d5529021818ee4774afb5412c).

feeley commented on doc/gambit.txi:
```diff
@@ -245,6 +246,122 @@ particular, the macOS and Microsoft Windows prebuilt installers set
+@end example
+
+If the code is in the scope of a @code{(declare-safe-define-procedure
+#t)} then it is vector-ref that is called instead of @code{##vector-ref}
```
Please add `@code{vector-ref}`.

feeley commented on doc/gambit.txi:
```diff
@@ -245,6 +246,122 @@ particular, the macOS and Microsoft Windows prebuilt installers set
+println to @code{##println} because those primitives perform the same
+operations (assuming the code has no errors). Note that foo and bar have
+remained the same (because they are not procedures exported by the
+runtime library) but pp has been mapped to @code{##pp#unimplemented}
```
Please add `@code{pp}`, also on next line.

feeley commented on doc/gambit.txi:
```diff
@@ -245,6 +246,122 @@ particular, the macOS and Microsoft Windows prebuilt installers set
+The namespace declarations in @code{~~lib/gambit/prim/prim#.scm} have
+caused a mapping of square to @code{##square}, car to @code{##car} and
+println to @code{##println} because those primitives perform the same
+operations (assuming the code has no errors). Note that foo and bar have
```
Please add `@code{foo}` and `@code{bar}`.

feeley commented on doc/gambit.txi:
```diff
@@ -245,6 +246,122 @@ particular, the macOS and Microsoft Windows prebuilt installers set
+@end example
+
+The namespace declarations in @code{~~lib/gambit/prim/prim#.scm} have
+caused a mapping of square to @code{##square}, car to @code{##car} and
```
Please add `@code{square}` and `@code{car}`.

feeley commented on doc/gambit.txi:
```diff
@@ -245,6 +246,122 @@ particular, the macOS and Microsoft Windows prebuilt installers set
+
+The @code{primitive} special form works in tandem with the
+@code{define-procedure} and @code{define-primitive} forms and the
+@code{~~lib/gambit/prim/prim#.scm} file and @code{gambit/prim} library.
```
Probably best to use the library name `(gambit prim)` to conform to R7RS.  Also a few lines below.

feeley commented on doc/gambit.txi:
```diff
@@ -245,6 +246,122 @@ particular, the macOS and Microsoft Windows prebuilt installers set
+remained the same (because they are not procedures exported by the
+runtime library) but pp has been mapped to @code{##pp#unimplemented}
+because pp is a procedure exported by the runtime library but
+@code{##pp} is not defined. Having unimplemented in the name will help
```
Use `@code{unimplemented}`.

feeley commented on doc/gambit.txi:
```diff
@@ -245,6 +246,122 @@ particular, the macOS and Microsoft Windows prebuilt installers set
+annotations in the parameter list and it inserts a @code{(include
+"~~lib/gambit/prim/prim#.scm")} in the body so that primitive operations
+can be used without the @code{##} prefix. Type checking and automatic
+forcing of promise arguments is also added implicitly. The form
```
is -> are

feeley commented on doc/gambit.txi:
```diff
@@ -245,6 +246,122 @@ particular, the macOS and Microsoft Windows prebuilt installers set
+with a @code{##} prefix, but other Scheme implementations could do it
+differently.
+
+Finally, there's the @code{(declare-safe-define-procedure <bool>)} that
```
that -> form that

feeley commented on doc/gambit.txi:
```diff
@@ -149,6 +149,7 @@ repository.
 
 @menu
 * Accessing the system files::      Accessing the system files
+* Extending the core system::       Extending the core system
```
This is too early in the documentation.  It is related to modules so should go in that chapter.

feeley commented on doc/gambit.txi:
```diff
@@ -245,6 +246,122 @@ particular, the macOS and Microsoft Windows prebuilt installers set
+
+If the code is in the scope of a @code{(declare-safe-define-procedure
+#t)} then it is vector-ref that is called instead of @code{##vector-ref}
+which will both check that x is a vector (redundantly) and that the
```
x - > `@code{x}`

feeley commented:
It would be good to add this table of type specifiers to the description... https://github.com/gambit/gambit/pull/541#issuecomment-622531223

lassik commented:
OK, fixed all of those. Added table and moved to Modules chapter as "Primitive Procedures" (maybe not the best name, but it's something to start with).
